From 5f7a291ff45374a85bf056c44ae1af2bed19f3ca Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Mon, 27 Nov 2023 10:25:54 +0800
Subject: [PATCH 3/9] tweak regex of ostree-system-generator

The ostree has boot params ostree="/ostree/1" on grub
and uboot which is a synlink point to
the dir contains "boot.0/OSTREE_OSNAME/***".

So tweak regex of ostree-system-generator.

Upstream-Status: Inappropriate [LAT specific]

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>

Rebased to ostree 2023.7

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>
---
 src/libostree/ostree-impl-system-generator.c | 7 ++++++-
 src/libostree/ostree-sysroot.c               | 2 +-
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/src/libostree/ostree-impl-system-generator.c b/src/libostree/ostree-impl-system-generator.c
index ad785eb..60b7103 100644
--- a/src/libostree/ostree-impl-system-generator.c
+++ b/src/libostree/ostree-impl-system-generator.c
@@ -144,7 +144,12 @@ fstab_generator (const char *ostree_cmdline, const char *normal_dir, const char
    * mounted yet.
    */
   g_autofree char *stateroot = NULL;
-  if (!_ostree_sysroot_parse_bootlink (ostree_cmdline, NULL, &stateroot, NULL, NULL, error))
+  char ostree_cmdline_new[PATH_MAX] = {0};
+
+  if (readlink(ostree_cmdline, ostree_cmdline_new, sizeof(ostree_cmdline_new)) < 0)
+      return glnx_prefix_error (error, "Failed to readlink %s", ostree_cmdline);
+
+  if (!_ostree_sysroot_parse_bootlink (ostree_cmdline_new, NULL, &stateroot, NULL, NULL, error))
     return glnx_prefix_error (error, "Parsing stateroot");
 
   /* Load /etc/fstab if it exists, and look for a /var mount */
diff --git a/src/libostree/ostree-sysroot.c b/src/libostree/ostree-sysroot.c
index 91b63f9..a7b0f39 100644
--- a/src/libostree/ostree-sysroot.c
+++ b/src/libostree/ostree-sysroot.c
@@ -730,7 +730,7 @@ _ostree_sysroot_parse_bootlink (const char *bootlink, int *out_entry_bootversion
   static GRegex *regex;
   if (g_once_init_enter (&regex_initialized))
     {
-      regex = g_regex_new ("^/ostree/boot.([01])/([^/]+)/([^/]+)/([0-9]+)$", 0, 0, NULL);
+      regex = g_regex_new ("boot.([01])/([^/]+)/([^/]+)/([0-9]+)$", 0, 0, NULL);
       g_assert (regex);
       g_once_init_leave (&regex_initialized, 1);
     }
-- 
2.27.0

